pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+        return False
+
+
+class MonomialTermExpression(ProductExpression):
+    __slots__ = ()
+
+
+class ReciprocalExpression(ExpressionBase):
+    """
+    Reciprocal expressions::
+
+        1/x
+    """
+    __slots__ = ()
+    PRECEDENCE = 3.5
+
+    def nargs(self):
+        return 1
+
+    def _precedence(self):
+        return ReciprocalExpression.PRECEDENCE
+
+    def _compute_polynomial_degree(self, result):
+        if result[0] == 0:
+            return 0
+        return None
+
+    def getname(self, *args, **kwds):
+        return 'recip'
+
+    def _to_string(self, values, verbose, smap, compute_values):
+        if verbose:
+            return "{0}({1})".format(self.getname(), values[0])
+        return "(1/{0})".format(values[0])
+
+    def _apply_operation(self, result):
+        return 1 / result[0]
+
+
+class NPV_ReciprocalExpression(ReciprocalExpression):
+    __slots__ = ()
+
+    def is_potentially_variable(self):
+        return False
+
+
+class _LinearOperatorExpression(ExpressionBase):
+    """
+    An 'abstract' class that defines the polynomial degree for a simple
+    linear operator
+    """
+
+    __slots__ = ()
+
+    def _compute_polynomial_degree(self, result):
+        # NB: We can't use max() here because None (non-polynomial)
+        # overrides a numeric value (and max() just ignores it)
+        ans = 0
+        for x in result:
+            if x is None:
+                return None
+            elif ans < x:
+                ans = x
+        return ans
+
+
+class RangedExpression(_LinearOperatorExpression):
+    """
+    Ranged expressions, which define relations with a lower and upper bound::
+
+        x < y < z
+        x <= y <= z
+
+    Args:
+        args (tuple): child nodes
+        strict (tuple): flags that indicates whether the inequalities are strict
+    """
+
+    __slots__ = ('_strict',)
+    PRECEDENCE = 9
+
+    def __init__(self, args, strict):
+        super(RangedExpression,self).__init__(args)
+        self._strict = strict
+
+    def nargs(self):
+        return 3
+
+    def construct_node(self, args, memo):
+        return self.__class__(args, self._strict)
+
+    def __getstate__(self):
+        state = super(RangedExpression, self).__getstate__()
+        for i in RangedExpression.__slots__:
+            state[i] = getattr(self, i)
+        return state
+
+    def __nonzero__(self):
+        return bool(self())
+
+    __bool__ = __nonzero__
+
+    def is_relational(self):
+        return True
+
+    def _precedence(self):
+        return RangedExpression.PRECEDENCE
+
+    def _apply_operation(self, result):
+        _l, _b, _r = result
+        if not self._strict[0]:
+            if not self._strict[1]:
+                return _l <= _b and _b <= _r
+            else:
+                return _l <= _b and _b < _r
+        elif not self._strict[1]:
+            return _l < _b and _b <= _r
+        else:
+            return _l < _b and _b < _r
+
+    def _to_string(sel
```

Comment:
I disagree with this note.  There is no reason for this to be any different from any other Unary Function